builtin/ui/style.lua

```diff
@@ -0,0 +1,262 @@
+		self._nested = table.merge(props.nested or props)
+		self._reset = props.reset
+
+		return self
+	end
+
+	if type(param) == "string" then
+		self._sel = ui._parse_sel(param)
+		return make_style
+	end
+
+	self._sel = ui._universal_sel
+	return make_style(param)
+end
+
+function ui.Style:_get_flat()
+	local flat_styles = {}
+	self:_get_flat_impl(flat_styles, ui._universal_sel)
+	return flat_styles
+end
+
+function ui.Style:_get_flat_impl(flat_styles, parent_sel)
+	-- Intersect our selector with our parent selector, resulting in a fully
+	-- qualified selector.
+	local full_sel = ui._intersect_sels({parent_sel, self._sel})
+
+	-- Copy this style's properties into a new style with the full selector.
+	local flat = ui.Style {
+		props = self._props,
+		reset = self._reset,
+	}
+	flat._sel = full_sel
+
+	table.insert(flat_styles, flat)
+
+	-- For each sub-style of this style, cascade it with our full selector and
+	-- add it to the list of flat styles.
+	for _, nested in ipairs(self._nested) do
+		nested:_get_flat_impl(flat_styles, full_sel)
+	end
+end
+
+local function cascade_layout(new, add, props)
+	new.layout = add.layout or props.layout
+	new.clip = add.clip or props.clip
+
+	new.scale = add.scale or props.scale
+end
+
+local function cascade_sizing(new, add, props)
+	new.size = add.size or props.size
+	new.span = add.span or props.span
+
+	new.pos = add.pos or props.pos
+	new.anchor = add.anchor or props.anchor
+
+	new.margin = add.margin or props.margin
+	new.padding = add.padding or props.padding
+end
+
+local function cascade_layer(new, add, props, p)
+	new[p.."_image"] = add[p.."_image"] or props[p.."_image"]
+	new[p.."_fill"] = add[p.."_fill"] or props[p.."_fill"]
+	new[p.."_tint"] = add[p.."_tint"] or props[p.."_tint"]
+
+	new[p.."_scale"] = add[p.."_scale"] or props[p.."_scale"]
+	new[p.."_source"] = add[p.."_source"] or props[p.."_source"]
+
+	new[p.."_frames"] = add[p.."_frames"] or props[p.."_frames"]
+	new[p.."_frame_time"] = add[p.."_frame_time"] or props[p.."_frame_time"]
+end
+
+function ui._cascade_props(add, props)
+	local new = {}
+
+	cascade_layout(new, add, props)
+	cascade_sizing(new, add, props)
+
+	new.display = add.display or props.display
+
+	cascade_layer(new, add, props, "box")
+	cascade_layer(new, add, props, "icon")
+
+	new.box_middle = add.box_middle or props.box_middle
+	new.box_tile = add.box_tile or props.box_tile
+
+	new.icon_place = add.icon_place or props.icon_place
+	new.icon_gutter = add.icon_gutter or props.icon_gutter
+	new.icon_overlap = add.icon_overlap or props.icon_overlap
+
+	return new
+end
+
+local function unpack_vec2d(vec)
+	if #vec == 2 then
+		return vec[1], vec[2]
+	elseif #vec == 1 then
+		return vec[1], vec[1]
+	end
+end
+
+local function unpack_rect(rect)
+	if #rect == 4 then
+		return rect[1], rect[2], rect[3], rect[4]
+	elseif #rect == 2 then
+		return rect[1], rect[2], rect[1], rect[2]
+	elseif #rect == 1 then
+		return rect[1], rect[1], rect[1], rect[1]
+	end
```

Comment:
Might want to make #rect a local `len` variable, rather than calling #rect 3 times.